### ISSUE TITLE
ErrorBoundary fails to catch error when passing props to component

### ISSUE THREAD

**Issue by ganzf** [facebook/react#13213]
**Do you want to request a *feature* or report a *bug*?**
Possibly both

**What is the current behavior?**
Currently, error boundaries will not catch an error when passing props to a component.
TL;DR : 
```
<ErrorBoundary>
    <SomeComponent evil={this.does.not.exist} />
</ErrorBoundary>
```

**Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**
I created a small public repository to show the current behavior.
You can find it here : https://github.com/ganzf/reactErrorBoundariesForInvalidProps

**What is the expected behavior?**
I wish to catch this kind of error. If there is some wrong code anywhere in the component rendering cycle (even when passing props) I wish to be able to control this error. In the current repository, the invalid code is obviously wrong and could be fixed by a developer. But what about more dynamic code such as `object[this.props.name]` ?

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
react 16.4.1
react-dom 16.4.1
react-scripts 1.1.4

**Comment by nmain** [comment]
It is caught, just at the next level up: https://jsfiddle.net/7v21wbmg/

I think this behavior makes perfect sense.  While `SomeComponent` may use `evil`, it doesn't know or care about `this.does.not.exist` at all.  The task of projecting from `this.does.not.exist` to `evil` is was something that `BadPonent` did in its render process, so it's an error in `BadPonent`, and like all error boundary errors, must be caught one level up.

**Comment by ganzf** [comment]
Hello and thank you for you answer ! 👍  

Alright, your explaination makes sense for this case.
What about this case : 

```// App.jsx
render() { 
...
return (
<ErrorBoundary>
    <SomeComponent />
</ErrorBoundary>
);

// SomeComponent.jsx
render() {
...
return (
<AnotherComponent evil={this.does.not.exist} />
);
```

Is this still the same use case ? If so, I'm not sure I understand the reasoning behind.

**Comment by ganzf** [comment]
I've spent the day tinkering, testing this and that. Something weird seems to happen.
I cannot expose the code directly nor make a working jsfiddle with react router, however I have noticed that the ErrorBoundary component will render the right view (ie `An error Occurred`), but after a few seconds the default development mode view (the ugly white and red screen with the stack trace). Checking the logs, the error is thrown only once, I enter the componentDidCatch function once, and I do not re-render the children or anything after the error has been caught.

I'm sorry that I can't provide more code right now. Thank you again for your time. 
I should mention I use react-router and redux, but the error is thrown in the render method of a component, so redux should not be of much concern..
